core/src/main/scala/kafka/server/metadata/BrokerMetadataListener.scala

```diff
@@ -205,6 +206,22 @@ class BrokerMetadataListener(
     put(MetadataLogEvent(messages, lastOffset))
   }
 
+  override def handleNewLeader(leader: MetaLogLeader): Unit = {
+    // Update MetadataCache w/ a new snapshot w/ the updated leader node ID
+    // Copy the existing snapshot
+    val snapshot = metadataCache.readState()
+
+    metadataCache.writeState(
+      MetadataSnapshot(
+        snapshot.partitionStates,
+        Some(leader.nodeId), // New leader ID
+        snapshot.aliveBrokers,
+        snapshot.aliveNodes,
+        snapshot.topicIdMap,
+        snapshot.fencedBrokers,
+        snapshot.brokerEpochs))
+  }
+
```

Comment:
There is concurrency problem here.  `PartitionMetadataProcessor` could be processing a batch of metadata records.  When it does so, it copies the current metadata cache state,  processes all of the records, and then writes the new metadata cache state.  If we write the state here, and that occurs after `PartitionMetadataProcessor` copied it but before it wrote out the new state based on its current batch, the state we write here will get overwritten.

Note that `PartitionMetadataProcessor` never changes the controllerId value in the metadata cache, so one possibility is to make it always take the current value -- i.e. replace all its occurrences of `mgr.getMetadataSnapshot().controllerId` with metadataCache.readState().controllerId`.

Reply:
There's a race condition even in this "quick solution".

Reply:
Spoke with Ron offline. I think it's easier to separate the ownership for updating the MetadataSnapshot fields to not introduce these race conditions. I'll go ahead and add APIs on the MetadataCache to let the `PartitionMetadataProcessor` update all fields except for the controller ID which will be owned by the `BrokerMetadataListener`.